### PR TITLE
enable vboost and set LM SM margin

### PR DESCRIPTION
**Pull request by malay-nagda**
# What does this PR do ?

enable vboost slider to boost performance

**Collection**: [llm]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by erhoo82 — APPROVED** [review]
LGTM

**Comment by maanug-nv** [comment]
This is just an example. If we want this behavior to be used by default (or easy to include in a custom recipe), we should create a NeMo Run plugin for this, correct? @hemildesai

**Comment by JimmyZhang12** [comment]
> This is just an example. If we want this behavior to be used by default (or easy to include in a custom recipe), we should create a NeMo Run plugin for this, correct? @hemildesai

Thanks Maanu, I refactored it into a NeMo-Run plugin

File: nemo/lightning/run/plugins.py [hunk 1]
```diff
@@ -241,3 +242,64 @@ def setup(self, task: run.Partial | run.Script, executor: run.Executor):
...
+            # Force program order kernel launch for TP, CP overlap
+            tp_size = task.trainer.strategy.tensor_model_parallel_size
+            cp_size = task.trainer.strategy.context_parallel_size
+            if tp_size > 1 and cp_size > 1:
+                executor.env_vars["CUDA_DEVICE_MAX_CONNECTIONS"] = 1
+
+            # Set LayerNorm SM margin when using P2P communication overlap to support the overlap with LayerNorm kernel
+            vpp = task.trainer.strategy.virtual_pipeline_model_parallel_size
+            pp_size = task.trainer.strategy.pipeline_model_parallel_size
+            if pp_size > 1 and vpp is not None and vpp > 1:
```
Comment (erhoo82) [inline nemo/lightning/run/plugins.py:290]: Don't we need to set this true even at PP=1 because we have DP comm overlap with LN here as well?

**Review by maanug-nv — DISMISSED** [review]

File: nemo/lightning/run/plugins.py (hunk 1 above)
Comment (jiemingz) [reply nemo/lightning/run/plugins.py]: Okay, Ive put this behind a knob which defaults to true.

File: nemo/lightning/run/plugins.py [hunk 2]
```diff
@@ -241,3 +242,65 @@ def setup(self, task: run.Partial | run.Script, executor: run.Executor):
...
+            # Force program order kernel launch for TP, CP overlap
+            tp_size = task.trainer.strategy.tensor_model_parallel_size
+            cp_size = task.trainer.strategy.context_parallel_size
+            if tp_size > 1 and cp_size > 1:
+                executor.env_vars["CUDA_DEVICE_MAX_CONNECTIONS"] = 1
+
+            # Set LayerNorm SM margin to support the overlap with LayerNorm kernel
+            if self.enable_layernorm_sm_margin:
+                executor.env_vars["NVTE_FWD_LAYERNORM_SM_MARGIN"] = 8
+                executor.env_vars["NVTE_BWD_LAYERNORM_SM_MARGIN"] = 8
```
Comment (erhoo82) [inline nemo/lightning/run/plugins.py:293]: Can we just make this to 16?
This should cover all NIC cases used for data-parallel collectives.

Comment (jiemingz) [reply nemo/lightning/run/plugins.py]: Okay, I also added a knob to specify the margin

**Review by erhoo82 — APPROVED** [review]
LGTM

**Pull request merged**